crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -80,32 +80,53 @@ impl Display for DisplayUnionType<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         let union = self.ty;
 
-        let (int_literals, other_types): (Vec<Type>, Vec<Type>) = union
+        // Group literals by type: {int, string, bytes, boolean, function, class}
+        let mut all_literals: Vec<Vec<String>> =
+            vec![vec![], vec![], vec![], vec![], vec![], vec![]];
+        let mut other_types = vec![];
+
+        union
             .elements(self.db)
             .iter()
             .copied()
-            .partition(|ty| matches!(ty, Type::IntLiteral(_)));
+            .for_each(|ty| match ty {
+                // Write a String for each literal type appended in order of appearance
+                Type::IntLiteral(n) => all_literals[0].push(format!("{n}")),
```

Comment:
Slightly tempted to suggest either using a struct instead of `all_literals`, or using constants instead of the magic number indices -- but I think this code is all close enough together that it's fine as-is. Using a struct would be significantly more code with arguable benefit, and using constants just doesn't seem necessary when all the indices are used once, in order, in close proximity.